### PR TITLE
OCC-25: SKUs should be globally unique

### LINKED ISSUE #107: SKUs should be globally unique (OCC-25)
Currently it's possible to create two content items with `ProductPart` that has the same SKU. SKUs need to be unique identifiers within the application so this causes problems.

Reproduction steps:
1. Create a new content type (I called it Offer) with `ProductPart` and `PricePart` to it.
2. Create a new content item, set SKU to "sample", give it a price, click Publish.
3. View the content item, click "Add to cart". It works.
4. Create a new content item, set SKU to "sample" again, give it a price, click Publish.
  - Expected: validation error for duplicate SKU
  - Actual: "Your Offer has been published. "
5. View either of the two content items and click  "Add to cart". You get the exception below:

```
ArgumentException: An item with the same key has already been added. Key: sample

    System.Collections.Generic.Dictionary<TKey, TValue>.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
    System.Collections.Generic.Dictionary<TKey, TValue>.Add(TKey key, TValue value)
    System.Linq.Enumerable.ToDictionary<TSource, TKey>(IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IEqualityComparer<TKey> comparer)
    System.Linq.Enumerable.ToDictionary<TSource, TKey>(IEnumerable<TSource> source, Func<TSource, TKey> keySelector)
    OrchardCore.Commerce.Services.PriceProvider.AddPrices(IEnumerable<ShoppingCartItem> items) in PriceProvider.cs
        var skuProducts = (await _productService.GetProducts(skus))
    OrchardCore.Commerce.Services.PriceService.AddPrices(IEnumerable<ShoppingCartItem> items) in PriceService.cs
        items = await priceProvider.AddPrices(items);
    OrchardCore.Commerce.Controllers.ShoppingCartController.AddItem(ShoppingCartLineUpdateModel line, string shoppingCartId) in ShoppingCartController.cs
        parsedLine = (await _priceService.AddPrices(new[] { parsedLine })).Single();
    lambda_method2280(Closure , object )
...
```

### PR DESCRIPTION
**Pull request by Psichorex**
OCC-25
Fixes #107

### CODE REVIEW

File: src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace OrchardCore.Commerce.Handlers;
+
+public class UniqueSkuValidationHandler : ContentPartHandler<ProductPart>
+{
+    private readonly ISession _session;
+    private readonly IUpdateModelAccessor _updateModelAccessor;
+
+    public UniqueSkuValidationHandler(
+        IContentManager contentManager,
```
Comment (DAud-IcI) [inline src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs:19]: Doesn't look like you need this.

File: src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        _session = session;
+        _updateModelAccessor = updateModelAccessor;
+    }
+
+    public override async Task UpdatedAsync(UpdateContentContext context, ProductPart instance)
+    {
+        var products = (await _session
+            .QueryContentItem(PublicationStatus.Published)
+            .ListAsync()).AsList();
```
Comment (DAud-IcI) [inline src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs:31]: You are downloading literally every content item in the DB. This will be slow (and possibly timeout) on any decently sized webapp, so please don't do that.

File: src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+        var isNotUnique = products
+            .Any(sku =>
+                sku.As<ProductPart>()?.Sku == instance.Sku &&
+                sku.ContentItem.ContentItemId != instance.ContentItem.ContentItemId);
```
Comment (DAud-IcI) [inline src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs:36]: Can't you do this in the query? Please check if there is an index you can use to filter this DB side. If not, then create it.

**Review by DAud-IcI — CHANGES_REQUESTED** [review]
Please also add a UI test, that creates a duplicate SKU and checks that the "SKU must be unique. A product with the given SKU already exists." model error is displayed.

File: src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs (hunk 2 above)
Comment (Psichorex) [reply src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs]: My problem is that I have to check the SKU because that has SKU but literally any ContentType there will be on the site using OCC could have ProductPart on it... Is there a way to tell the DB I only want to query those ContentItems that have ProductPart on them? Because I have never done such query before.

Comment (DAud-IcI) [reply src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs]: We already have [`ProductPartindex`](https://github.com/OrchardCMS/OrchardCore.Commerce/blob/main/src/Modules/OrchardCore.Commerce/Indexes/ProductPartIndex.cs). It contains properties for both ContentItem and Sku too. Use `.With<TIndex>()` to filter by it.

Comment (Psichorex) [reply src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs]: I am experimenting with writing a Query instead of QueryContentItem I will update here

Comment (DAud-IcI) [reply src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs]: Why? `session.QueryContentItem(PublicationStatus.Published)` is just `session.Query<ContentItem, ContentItemIndex>(index => index.Published)`.

Comment (Psichorex) [reply src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs]: `Query<ProductPart, ProductPartIndex>(index => index.Sku == instance.Sku && index.ContentItemId != instance.ContentItem.ContentItemId)` this yields no result although it really should because the criteria is met I am checking in DB browser. I don't know why but if I can't DB side check SKUs I won't be able to omit querying all the contentitems.

Comment (DAud-IcI) [reply src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs]: That's not acceptable, because as I said it will timeout on a bigger site.

`Query<ProductPart, ProductPartIndex>` should be `Query<ContentItem, ProductPartIndex>`.

Comment (Psichorex) [reply src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs]: Same goes for 
`Query().For<ProductPart>().With<ProductPartIndex>(index => index.Sku == instance.Sku)` returns null altough there is like 5 contentitems with the given SKU already

Comment (Psichorex) [reply src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs]: okay the problem was refering to  ProductPart

Comment (DAud-IcI) [reply src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs]: Again, you are supposed to query ContentItem, not ProductPart.

Comment (Psichorex) [reply src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs]: Yes I know now, just your messages were coming with a Delay so I already wrote my question when I saw your answer

File: src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs [hunk 4]
```diff
@@ -26,16 +25,15 @@ public class UniqueSkuValidationHandler : ContentPartHandler<ProductPart>
...
     public override async Task UpdatedAsync(UpdateContentContext context, ProductPart instance)
     {
-        var products = (await _session
-            .QueryContentItem(PublicationStatus.Published)
-            .ListAsync()).AsList();
+        var isProductSkuAlreadyExisting = (await _session
+                .Query<ContentItem, ProductPartIndex>(index =>
+                    index.Sku == instance.Sku &&
+                    index.ContentItemId != instance.ContentItem.ContentItemId)
+                .ListAsync())
```
Comment (DAud-IcI) [inline src/Modules/OrchardCore.Commerce/Handlers/UniqueSkuValidationHandler.cs:32]: If you use `.CountAsync()` then it doesn't have to download the results either.

**Review by DAud-IcI — CHANGES_REQUESTED** [review]
